File: loopx/control_plane/todos/succession.py
"""Normalize compatibility input for the one typed succession read policy."""
from __future__ import annotations

from typing import Any

from ..effect_runtime import effect_runtime_result
from .contract import (
    normalize_todo_id, normalize_todo_id_list, normalize_todo_status,
    normalize_todo_task_class, normalize_todo_no_followup,
    normalize_todo_resume_when, normalize_todo_excluded_agents,
)


def succession_facts(item: dict[str, Any]) -> dict[str, Any]:
    resume = normalize_todo_resume_when(item.get("resume_when")) or ""
    return {
        "todo_id": normalize_todo_id(item.get("todo_id")),
        "status": normalize_todo_status(item.get("status")) or ("done" if item.get("done") else "open"),
        "active": (item.get("archive_state") or "active") == "active",
        "advancement": normalize_todo_task_class(item.get("task_class"), text=str(item.get("text") or ""),
            action_kind=item.get("action_kind")) == "advancement_task",
        "no_followup": normalize_todo_no_followup(item.get("no_followup")) is True,
        "successors": normalize_todo_id_list(item.get("successor_todo_ids")),
        "superseded_by": normalize_todo_id(item.get("superseded_by")),
        "unblocks": normalize_todo_id(item.get("unblocks_todo_id")),
        "resumes": normalize_todo_id(resume.partition(":")[2]) if resume.startswith("todo_done:") else None,
        "handoff": bool(normalize_todo_excluded_agents(item.get("excluded_agents")))
            and bool(normalize_todo_id(item.get("unblocks_todo_id"))),
        "context_fields": sorted(key for key, value in item.items()
            if value is not None and key != "succession_evaluation"),
    }


def project_succession(items: list[dict[str, Any]], *, reuse: bool = False) -> list[dict[str, Any]]:
    rows = [succession_facts(item) for item in items]
    # Archived histories repeat field-presence shapes thousands of times. Intern
    # those shapes, preserving the complete graph inside the existing RPC budget.
    contexts: list[list[str]] = []
    context_ids: dict[tuple[str, ...], int] = {}
    for row in rows:
        shape = tuple(row["context_fields"])
        if shape not in context_ids:
            context_ids[shape] = len(contexts)
            contexts.append(list(shape))
        row["context_fields"] = context_ids[shape]
    request: dict[str, Any] = {"schema_version": "todo_succession_request_v0",
        "rows": rows, "context_field_sets": contexts}
    if reuse:
        request["evaluations"] = [item.get("succession_evaluation") for item in items]
    result = effect_runtime_result("todo.succession.project", request)
    if not isinstance(result, dict) or result.get("schema_version") != "todo_succession_result_v0":
        raise ValueError("invalid typed Todo succession result")
    evaluations = result.get("evaluations")
    if not isinstance(evaluations, list) or len(evaluations) != len(items) or any(not isinstance(row, dict) for row in evaluations):
        raise ValueError("invalid typed Todo succession cardinality")
    return evaluations


def evaluate_succession(items: list[dict[str, Any]], lineage: list[dict[str, Any]] | None = None) -> None:
    # Active evaluated rows override their unevaluated source versions; retained
    # history and other roles remain graph evidence but never become active rows.
    if not items:
        return
    # Replace one matching source row, not every occurrence of its identity.
    # Duplicate archive/role identities must remain visible to the TS validator.
    selected = {(normalize_todo_id(item.get("todo_id")), item.get("role"), item.get("archive_state") or "active")
        for item in items}
    source = []
    for item in lineage or []:
        key = (normalize_todo_id(item.get("todo_id")), item.get("role"), item.get("archive_state") or "active")
        if key in selected:
            selected.remove(key)
        else:
            source.append(item)
    source.extend(items)
    evaluations = project_succession(source)
    for item, evaluation in zip(items, evaluations[len(source) - len(items):], strict=True):
        item["succession_evaluation"] = evaluation


def public_todo_summary(summary: dict[str, Any]) -> dict[str, Any]:
    """Drop the internal full-graph handoff once a consumer has selected rows."""
    return {**summary, "items": [
        {key: value for key, value in item.items() if key != "succession_evaluation"}
        if isinstance(item, dict) else item
        for item in summary.get("items") or []
    ]}

File: loopx/control_plane/todos/succession.ts
/** Read-only continuation evidence over the complete Todo graph.
 * These evaluations are derived display state, never completion or execution authority. */
import {canonicalAuthoritySha256} from "../coordination/authority_store_codec.ts";
import {EffectRuntimeRequestError} from "../effect_runtime_errors.ts";
import type {JsonObject} from "../effect_program.ts";
import {requireBoolean, requireJsonObject, requireStringLiteral} from "../runtime_decode.ts";

export const HANDOFF_STATES = ["blocking", "cleared_without_successor", "cleared_with_successor",
  "cleared_no_followup", "superseded", "deferred"] as const;
export type HandoffState = typeof HANDOFF_STATES[number];
const CONTEXT_FIELDS = ["action_kind", "task_repository", "continuation_policy", "claimed_by",
  "completed_at", "updated_at", "required_write_scopes", "required_capabilities", "target_capabilities",
  "explore_result_node_refs", "decision_scope", "required_decision_scopes", "unblocks_todo_id",
  "resume_when", "blocks_agent", "excluded_agents", "global_gate"] as const;
const EVALUATION_SCHEMA = "todo_succession_evaluation_v0";
interface Row {
  facts: JsonObject; id: string | null; status: "open" | "blocked" | "done" | "deferred";
  active: boolean; advancement: boolean; noFollowup: boolean; tracked: boolean;
  successors: string[]; supersededBy: string | null; unblocks: string | null;
  resumes: string | null; handoff: boolean;
}
function id(value: unknown): string | null {
  if (value === null) return null;
  if (typeof value !== "string" || !/^todo_[a-z0-9_-]{3,64}$/.test(value)) {
    throw new EffectRuntimeRequestError("succession requires normalized Todo identities");
  }
  return value;
}
function decode(value: unknown): Row {
  const raw = requireJsonObject(value, "succession facts");
  const facts: JsonObject = {...raw, context_fields: Array.isArray(raw.context_fields)
    ? CONTEXT_FIELDS.filter(field => (raw.context_fields as unknown[]).includes(field)) : raw.context_fields};
  const status = requireStringLiteral(facts.status, ["open", "blocked", "done", "deferred"], "status");
  const active = requireBoolean(facts.active, "active");
  const advancement = requireBoolean(facts.advancement, "advancement");
  if (!Array.isArray(facts.successors) || !Array.isArray(facts.context_fields) ||
      facts.context_fields.some(field => typeof field !== "string")) {
    throw new EffectRuntimeRequestError("succession lists must be arrays");
  }
  return {facts, id: id(facts.todo_id), status, active, advancement,
    noFollowup: requireBoolean(facts.no_followup, "no_followup"),
    tracked: active && status === "done" && advancement &&
      CONTEXT_FIELDS.some(field => (facts.context_fields as unknown[]).includes(field)),
    successors: facts.successors.map(value => {
      const target = id(value);
      if (!target) throw new EffectRuntimeRequestError("successor identity cannot be null");
      return target;
    }), supersededBy: id(facts.superseded_by), unblocks: id(facts.unblocks),
    resumes: id(facts.resumes), handoff: requireBoolean(facts.handoff, "handoff")};
}
function handoffState(row: Row, successors: readonly string[]): HandoffState | null {
  if (!row.active || !row.handoff) return null;
  if (row.supersededBy && successors.includes(row.supersededBy)) return "superseded";
  if (row.status === "deferred") return "deferred";
  if (row.status !== "done") return "blocking";
  if (row.noFollowup) return "cleared_no_followup";
  return successors.length ? "cleared_with_successor" : "cleared_without_successor";
}

/** The same edge index drives live readback and bounded archive capture. */
export function indexInferredSuccessors(rows: readonly Pick<Row, "id" | "advancement" | "unblocks" | "resumes">[]): Map<string, string[]> {
  const inferred = new Map<string, string[]>();
  for (const row of rows) {
    if (!row.id || !row.advancement) continue;
    for (const source of new Set([row.unblocks, row.resumes])) {
      if (source && source !== row.id) {
        const targets = inferred.get(source) ?? [];
        targets.push(row.id);
        inferred.set(source, targets);
      }
    }
  }
  return inferred;
}

/** Index inferred edges once; both completion and handoff use the same resolver. */
export function evaluateTodoSuccession(values: readonly unknown[]): JsonObject[] {
  const rows = values.map(decode), byId = new Map<string, Row>();
  for (const row of rows) {
    if (!row.id) continue;
    if (byId.has(row.id)) throw new EffectRuntimeRequestError(`duplicate succession identity: ${row.id}`);
    byId.set(row.id, row);
  }
  const inferred = indexInferredSuccessors(rows);
  return rows.map(row => {
    const declared = [...new Set([...row.successors, ...(row.supersededBy ? [row.supersededBy] : [])])];
    // A retained archived target is still evidence. A missing/self target is not.
    const resolved = declared.filter(target => target !== row.id && byId.has(target));
    const successors = [...new Set([...resolved, ...(row.id ? inferred.get(row.id) ?? [] : [])])];
    return {schema_version: EVALUATION_SCHEMA, item_sha256: canonicalAuthoritySha256(row.facts),
      successor_todo_ids: successors, unresolved_successor_ids: declared.filter(target => !resolved.includes(target)),
      tracked_completion: row.tracked, successor_gap: row.tracked && !row.noFollowup && successors.length === 0,
      handoff_state: handoffState(row, successors)};
  });
}

/** Filtering may reuse a fresh full-source result, but may not change its item facts. */
export function validateTodoSuccession(facts: unknown, value: unknown): JsonObject {
  const row = decode(facts), evaluation = requireJsonObject(value, "succession evaluation");
  if (evaluation.schema_version !== EVALUATION_SCHEMA || evaluation.item_sha256 !== canonicalAuthoritySha256(row.facts) ||
      !Array.isArray(evaluation.successor_todo_ids) || !Array.isArray(evaluation.unresolved_successor_ids)) {
    throw new EffectRuntimeRequestError("Todo display requires a matching full-source succession evaluation");
  }
  for (const target of [...evaluation.successor_todo_ids, ...evaluation.unresolved_successor_ids]) {
    if (id(target) === null) throw new EffectRuntimeRequestError("successor identity cannot be null");
  }
  requireBoolean(evaluation.tracked_completion, "tracked_completion");
  requireBoolean(evaluation.successor_gap, "successor_gap");
  if (evaluation.handoff_state !== null) requireStringLiteral(evaluation.handoff_state, HANDOFF_STATES, "handoff_state");
  const successors = evaluation.successor_todo_ids as string[];
  if (evaluation.tracked_completion !== row.tracked ||
      evaluation.successor_gap !== (row.tracked && !row.noFollowup && successors.length === 0) ||
      evaluation.handoff_state !== handoffState(row, successors) ||
      new Set(successors).size !== successors.length || successors.includes(row.id ?? "")) {
    throw new EffectRuntimeRequestError("inconsistent Todo succession evaluation");
  }
  return evaluation;
}

export function projectTodoSuccession(value: unknown): JsonObject {
  const request = requireJsonObject(value, "Todo succession request");
  if (request.schema_version !== "todo_succession_request_v0" || !Array.isArray(request.rows)) {
    throw new EffectRuntimeRequestError("Todo succession request schema mismatch");
  }
  const contexts = request.context_field_sets;
  if (contexts !== undefined && (!Array.isArray(contexts) || contexts.some(value =>
      !Array.isArray(value) || value.some(field => typeof field !== "string")))) {
    throw new EffectRuntimeRequestError("invalid succession context field sets");
  }
  const rows = request.rows.map(value => {
    const row = requireJsonObject(value, "succession row");
    if (!Array.isArray(contexts)) return row;
    const index = row.context_fields;
    if (typeof index !== "number" || !Number.isSafeInteger(index) || index < 0 || index >= contexts.length) {
      throw new EffectRuntimeRequestError("invalid succession context field index");
    }
    return {...row, context_fields: contexts[index]};
  });
  const evaluations = request.evaluations;
  if (evaluations !== undefined && (!Array.isArray(evaluations) || evaluations.length !== request.rows.length)) {
    throw new EffectRuntimeRequestError("succession evaluation cardinality mismatch");
  }
  return {schema_version: "todo_succession_result_v0", evaluations: Array.isArray(evaluations)
    ? rows.map((row, index) => validateTodoSuccession(row, evaluations[index]))
    : evaluateTodoSuccession(rows)};
}

/** Summary proofs are derived from every selected row before display caps.
 * A query subset can describe items but cannot certify closure of the source. */
export function projectTodoClosure(value: unknown): JsonObject {
  const request = requireJsonObject(value, "Todo closure request");
  if (request.schema_version !== "todo_closure_request_v0" || !Array.isArray(request.rows)) {
    throw new EffectRuntimeRequestError("Todo closure request schema mismatch");
  }
  const source = typeof request.source_section === "string" ? request.source_section : "";
  const role = request.role;
  const valid = (role === "user" || role === "agent") && source.trim() !== "" &&
    requireBoolean(request.full_selection, "full_selection");
  const rows = request.rows.map(value => {
    const row = requireJsonObject(value, "closure row");
    return {status: requireStringLiteral(row.status, ["open", "blocked", "done", "deferred"], "status"),
      watch: requireBoolean(row.watch_only, "watch_only"), noFollowup: requireBoolean(row.no_followup, "no_followup"),
      gap: requireBoolean(row.successor_gap, "successor_gap"), replan: requireBoolean(row.replan, "replan"),
      handoff: row.handoff_state === null ? null : requireStringLiteral(row.handoff_state, HANDOFF_STATES, "handoff_state")};
  });
  const noFollowup = rows.filter(row => (row.status === "done" || row.status === "deferred") && row.noFollowup).length;
  const watches = rows.filter(row => row.status !== "done" && row.status !== "deferred" && row.watch).length;
  const convergent = rows.filter(row => row.status !== "done" && row.status !== "deferred" && !row.watch).length;
  const result: JsonObject = {};
  if (valid && convergent === 0 && rows.every(row => row.status !== "deferred")) {
    result.source_proof = {schema_version: "todo_source_proof_v0", role, item_count: rows.length, derived: true};
  }
  if (valid && rows.every(row => (row.status === "done" || row.watch) && row.status !== "deferred" &&
      !row.gap && !row.replan && row.handoff !== "cleared_without_successor" && row.handoff !== "blocking")) {
    result.terminal_closure_proof = {schema_version: "todo_terminal_closure_proof_v0", role,
      source_section: source, item_count: rows.length, all_todos_done: watches === 0,
      monitor_open_count: watches, successor_gap_count: 0, route_replan_count: 0,
      no_followup_count: noFollowup, derived: true,
      ...(watches ? {all_convergent_todos_done: true, watch_only_monitor_count: watches} : {})};
  }
  if (noFollowup) result.closure_intent = {schema_version: "todo_closure_intent_v0", kind: "no_followup", derived: true, count: noFollowup};
  return result;
}
